File: ical_gen.py
from icalendar import Calendar, Event, Timezone, TimezoneStandard, TimezoneDaylight
from datetime import datetime
import uuid
import pytz
from models import VrbneEvent

TZ = pytz.timezone("Europe/Prague")

def create_ical(events: list[VrbneEvent], include_weather: bool = True) -> str:
    cal = Calendar()
    cal.add('prodid', '-//vrbne-docker//NONSGML v1.0//EN')
    cal.add('version', '2.0')
    cal.add('x-wr-calname', 'Kanál České Vrbné')
    cal.add('x-wr-timezone', 'Europe/Prague')

    # Přidání VTIMEZONE pro lepší kompatibilitu (Outlook)
    tz_component = create_timezone_component()
    cal.add_component(tz_component)

    for event in events:
        ical_event = Event()
        ical_event.add('uid', f"{uuid.uuid4()}@vrbne-docker")
        ical_event.add('dtstamp', datetime.now(pytz.utc))
        ical_event.add('dtstart', event.datetime_from(TZ))
        ical_event.add('dtend', event.datetime_to(TZ))
        
        summary = "🌊 Vrbné"
        if include_weather and event.weather:
            summary += f" {get_weather_icon(event.weather.icon)}{int(event.weather.feels_like)}°C"
        if event.note:
            summary += f" - {event.note}"
        
        ical_event.add('summary', summary)
        
        description = "Provoz slalomového kanálu\nZdroj: http://itdev.cz/SlalomCourse/OpeningTimes.aspx"
        if include_weather and event.weather:
            w = event.weather
            description += f"\n\nPočasí ({w.date.strftime('%H:%M')}):\n"
            description += f"🌡️ Teplota: {w.temp}°C (pocitově {w.feels_like}°C)\n"
            description += f"🌬️ Vítr: {w.wind_speed} m/s\n"
            description += f"☁️ {w.desc}"
            
        ical_event.add('description', description)
        cal.add_component(ical_event)

    return cal.to_ical().decode('utf-8')

def get_weather_icon(icon_code: str) -> str:
    icons = {
        "01d": "☀️", "01n": "🌕",
        "02d": "⛅", "02n": "🌤️",
        "03d": "☁️", "03n": "☁️",
        "04d": "☁️", "04n": "☁️",
        "09d": "🌧️", "09n": "🌧️",
        "10d": "🌦️", "10n": "🌦️",
        "11d": "⛈️", "11n": "⛈️",
        "13d": "❄️", "13n": "❄️",
        "50d": "🌫️", "50n": "🌫️",
    }
    return icons.get(icon_code, "")

def create_timezone_component():
    # Vytvoření standardní VTIMEZONE komponenty pro Europe/Prague
    tz = Timezone()
    tz.add('tzid', 'Europe/Prague')
    
    # Zjednodušeně pro ČR (střídání času)
    # V praxi by bylo lepší použít knihovnu, ale icalendar vyžaduje ruční plnění
    # Pro účely tohoto úkolu stačí základní definice
    
    std = TimezoneStandard()
    std.add('tzname', 'CET')
    std.add('dtstart', datetime(1970, 10, 25, 3, 0, 0))
    std.add('rrule', {'freq': 'yearly', 'bymonth': 10, 'byday': ['-1su']})
    std.add('tzoffsetfrom', pytz.timezone("Europe/Prague").utcoffset(datetime(2023, 6, 1))) # +2
    std.add('tzoffsetto', pytz.timezone("Europe/Prague").utcoffset(datetime(2023, 1, 1)))   # +1
    
    dst = TimezoneDaylight()
    dst.add('tzname', 'CEST')
    dst.add('dtstart', datetime(1970, 3, 29, 2, 0, 0))
    dst.add('rrule', {'freq': 'yearly', 'bymonth': 3, 'byday': ['-1su']})
    dst.add('tzoffsetfrom', pytz.timezone("Europe/Prague").utcoffset(datetime(2023, 1, 1))) # +1
    dst.add('tzoffsetto', pytz.timezone("Europe/Prague").utcoffset(datetime(2023, 6, 1)))   # +2
    
    tz.add_component(std)
    tz.add_component(dst)
    return tz

File: src/ical_gen.ts
import crypto from 'crypto';
import IcalGenerator from 'ical-generator';
import { DateTime } from 'luxon';
import { VrbneEvent } from './models.js';

const TZ = 'Europe/Prague';

export function createIcal(events: VrbneEvent[], includeWeather: boolean = true): string {
  const cal = new (IcalGenerator as any)({
    prodId: '-//vrbne-docker//NONSGML v1.0//EN',
    name: 'Kanál České Vrbné',
    timezone: TZ,
  });

  for (const event of events) {
    let summary = '🌊 Vrbné';

    if (includeWeather && event.weather) {
      const weatherIcon = getWeatherIcon(event.weather.icon);
      summary += ` ${weatherIcon}${Math.round(event.weather.feels_like)}°C`;
    }

    if (event.note) {
      summary += ` - ${event.note}`;
    }

    const startDt = DateTime.fromISO(
      `${event.date}T${event.time_from}:00`,
      { zone: TZ }
    );
    const endDt = DateTime.fromISO(
      `${event.date}T${event.time_to}:00`,
      { zone: TZ }
    );

    let description = 'Provoz slalomového kanálu\nZdroj: http://itdev.cz/SlalomCourse/OpeningTimes.aspx';

    if (includeWeather && event.weather) {
      const w = event.weather;
      description += `\n\nPočasí (${w.date.toFormat('HH:mm')}):\n`;
      description += `🌡️ Teplota: ${w.temp}°C (pocitově ${w.feels_like}°C)\n`;
      description += `🌬️ Vítr: ${w.wind_speed} m/s\n`;
      description += `☁️ ${w.desc}`;
    }

    // Generate a deterministic ID based on date and time
    const eventIdSource = `${event.date}-${event.time_from}-${event.time_to}`;
    const eventId = crypto.createHash('sha1').update(eventIdSource).digest('hex');

    const eventObj = cal.createEvent({
      id: eventId,
      summary,
      description,
      start: startDt,
      end: endDt,
      // Use the start date of the event as the timestamp (DTSTAMP)
      // to avoid triggering updates in calendar apps just because the generation time changed.
      timestamp: startDt.toJSDate(),
    });
    eventObj.timezone(TZ);
  }

  return cal.toString();
}

function getWeatherIcon(iconCode: string): string {
  const icons: Record<string, string> = {
    '01d': '☀️',
    '01n': '🌕',
    '02d': '⛅',
    '02n': '🌤️',
    '03d': '☁️',
    '03n': '☁️',
    '04d': '☁️',
    '04n': '☁️',
    '09d': '🌧️',
    '09n': '🌧️',
    '10d': '🌦️',
    '10n': '🌦️',
    '11d': '⛈️',
    '11n': '⛈️',
    '13d': '❄️',
    '13n': '❄️',
    '50d': '🌫️',
    '50n': '🌫️',
  };

  return icons[iconCode] || '';
}
